deepinv/transform/projective.py

```diff
@@ -0,0 +1,330 @@
+    :return torch.Tensor | Image.Image: transformed image.
+    """
+
+    assert interpolation in ("bilinear", "bicubic", "nearest")
+
+    w, h = (im.shape[2], im.shape[3]) if isinstance(im, torch.Tensor) else im.size
+    u0, v0 = int(w / 2), int(h / 2)
+    f = 100
+    s = 0
+    m_x = m_y = 1
+
+    # fmt: off
+    K = np.array([
+        [f*m_x, s, u0],
+        [0, f*m_y, v0],
+        [0, 0, 1]
+    ])
+
+    K_dash = np.array([
+        [f/zoom_factor*m_x/x_stretch_factor, s + skew, u0 + x_t],
+        [0, f/zoom_factor*m_y/y_stretch_factor, v0 + y_t],
+        [0, 0, 1]
+    ])
+    # fmt: on
+
+    R_dash = Rotation.from_euler(
+        "xyz", [theta_x, theta_y, theta_z], degrees=True
+    ).as_matrix()
+
+    if isinstance(im, torch.Tensor):
+        # note thetas defined in the opposite direction here, but it doesn't matter
+        # for random transformations which have symmetric ranges about 0.
+        H_inverse = K @ R_dash @ np.linalg.inv(K_dash)
+
+        if verbose:
+            with np.printoptions(precision=2, suppress=True):
+                print(H_inverse)
+
+        return warp_perspective(
+            im.double(),
+            torch.from_numpy(H_inverse)[None].to(device),
+            dsize=im.shape[2:],
+            mode=interpolation,
+            padding_mode=padding,
+        )
+    else:
+
+        if interpolation == "bilinear":
+            pil_interp = Image.Resampling.BILINEAR
+        elif interpolation == "bicubic":
+            pil_interp = Image.Resampling.BICUBIC
+        elif interpolation == "nearest":
+            pil_interp = Image.Resampling.NEAREST
+
+        H = K_dash @ R_dash @ np.linalg.inv(K)
+
+        return im.transform(
+            size=(im.size[0], im.size[1]),
+            method=Image.Transform.PERSPECTIVE,
+            data=H.flatten(),
+            resample=pil_interp,
+        )
+
+
+@dataclass
+class Homography(torch.nn.Module):
+    """
+    Homography (or projective transformation).
+
+    The homography is parameterised by
+    geometric parameters. By fixing these parameters, subgroup transformations are
+    retrieved, see Wang et al. "Perspective-Equivariant Imaging: an Unsupervised
+    Framework for Multispectral Pansharpening" https://arxiv.org/abs/2403.09327
+
+    For example, setting x_stretch_factor_min = y_stretch_factor_min = zoom_factor_min = 1,
+    theta_max = theta_z_max = skew_max = 0 gives a pure translation.
+
+    Subgroup transformations include :class:`deepinv.transform.Affine`, :class:`deepinv.transform.Similarity`,
+    :class:`deepinv.transform.Euclidean` along with the basic :class:`deepinv.transform.Shift`,
+    :class:`deepinv.transform.Rotation` and semigroup :class:`deepinv.transform.Scale`.
+
+    Transformations with perspective effects (i.e. pan+tilt) are recovered by setting
+    theta_max > 0.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
```

Comment:
the standard way of adding examples is adding first a separator via |sep| 
and then using >>> (otherwise the doctests are not run,  and the automatic copy-paste button will not work) as in [here](https://github.com/deepinv/deepinv/blob/0d0c174f4e5fa9031439bcac82631420c745ed28/deepinv/physics/blur.py#L323)